Add fast AV1 decoding/source file support via dav1d

f3ndot opened the pull request:
**Description of Change:**

This simply attempts to compile ffmpeg with libdav1d support that's cross-compile friendly.

Fixes #1737, partially solves #457 (decode).

I chose libdav1d as its the _de facto_ decoder which is faster than the reference library, libaom. It appears to be the horse everyone is betting on and is financially supported by the Alliance for Open Media/AOM.

It also appears to be more stable, using various Netflix test AV1 files, it doesn't hang on some of them like libaom does. And is of course faster.

~~In order to compile ffmpeg with libdav1d support, I needed to get on the nightly release. I chose the latest master snapshot and removed the patches that were merged in the upstream.~~ With ffmpeg 4.2 released, libdav1d was supported officially.

It uses the `meson` build system though, which means that's a new package requirement for compiling Handbrake from source. If the maintainers approve this, I'll add the appropriate HandBrake-docs PR.

**Test on:**

- [x] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux

f3ndot committed:
contrib: Use libdav1d for AV1 decoding in FFmpeg.

EwoutH commented:
Very nice work, good to see dav1d get adopted.

It's indeed faster on content most users use, but not (yet) on all systems and all content. See [dav1d 0.1.0 release: The first benchmarks](https://medium.com/@ewoutterhoeven/dav1d-0-1-0-release-the-first-benchmarks-5404360e44e3).

f3ndot commented:
Fighting with Travis' meson version. It all works locally on my Ubuntu 18.10 (Cosmic) setup. Once I get the build to pass I think this is ready for merge

f3ndot commented:
Travis' build system has been changed from xenial to bionic in https://github.com/HandBrake/HandBrake/pull/1864/commits/c735ff0b5d17323f889a76f72e1e842ec7b0017e since libdav1d needs a newer version of `nasm` than what is available in xenial.

Since the HandBrake documentation currently expects the build system to be Bionic anyway, I figure this is a safe and acceptable thing to do.

**EDIT** I'm a fool. Bionic is not yet supported in Travis :disappointed:

f3ndot commented:
Score! libdav1d is properly building in Travis now. The libvpx failure is unrelated and is addressed in #1867 

@sr55 would love review and merge of this when you have time

sr55 commented:
@jstebbins @bradleysepos @galad87 

Are you guys happy with this?

jstebbins reviewed (commented):
I can't speak for the travis changes.

It uses an non-release version of ffmpeg which will probably cause failures in all distro builds of HandBrake since they like to substitute their system libs which will be the last release version.

It adds 2 new build dependencies, ninja and meson.  So you'll have to update build machines.  I assume these can be compiled on OSX?

Aside from the above, it builds fine for me.  I don't have any AV1 source files handy to actually test it though. 

If there are no build issues on the other platforms, I think it's fine for nightly builds.  I would prefer that we use a release version of ffmpeg for HandBrake releases, but that's not a deal killer either.

sr55 commented:
We could gate it until the code required is in a public ffmpeg build.

bradleysepos commented:
True. Frankly, I'm not for releasing HandBrake against FFmpeg master, especially to support a currently not widespread codec. A lot of bug fixes go into their release branches, and I think that's important.

bradleysepos commented:
Perhaps it would be better to maintain this as an official feature branch (e.g. HandBrake/libdav1d) until we can be confident we're not giving up a large part of the upstream testing and bug fixing process, and then merge to master in prep for the next HandBrake release at that time.

f3ndot commented:
> It adds 2 new build dependencies, ninja and meson. So you'll have to update build machines. I assume these can be compiled on OSX?

Yes, it compiles on MacOS

> I don't have any AV1 source files handy to actually test it though.

Netflix makes available AV1 test files here: http://download.opencontent.netflix.com/?prefix=AV1/Chimera/

>  I'm not for releasing HandBrake against FFmpeg master

This makes sense. I'm happy with whatever y'all wish to do (feature branch seems nice, with maybe some "unstable" libdav1d-supported bins available on HB's website)

EwoutH commented:
dav1d 0.2.0 will get released this week with a lot of speed ups for SSSE3 CPU's (older ones without AVX2 support) on 8-bit content.

Nomis101 commented:
> dav1d 0.2.0 will get released this week

Now it is. :-)
https://code.videolan.org/videolan/dav1d/releases

f3ndot commented:
Safe to say this PR is parked until ffmpeg releases 4.2 with libdav1d support. At which time i'll update the PR for the latest lib.

EwoutH commented:
dav1d 0.2.1 is [tagged](https://code.videolan.org/videolan/dav1d/tags/0.2.1). FFmpeg will be updated soon.

EwoutH commented:
[dav1d 0.3.0](https://code.videolan.org/videolan/dav1d/tags/0.3.0) is released.

My average numbers (compared to 0.2.1)
- +24% performance on SSSE3
- +26% performance on SSE4.1
- +4% performance on AVX2
- +12 performance on Arm64

Nomis101 commented:
> [dav1d 0.3.0](https://code.videolan.org/videolan/dav1d/tags/0.3.0) is released.
> 
> My average numbers (compared to 0.2.1)

Do you see additional performance improvements with 0.4.0 'Cheetah'?

Nomis101 commented:
> Safe to say this PR is parked until ffmpeg releases 4.2 with libdav1d support. 

FFmpeg 4.2 is released now :-)

sr55 commented:
We'll hopefully get ffmpeg 4.2 in soon and we can get this rebased and committed if all looks good.

bradleysepos commented:
https://github.com/HandBrake/HandBrake/pull/2234

sr55 commented on .circleci/config.yml:
```diff
@@ -0,0 +1,210 @@
+version: 2
```
Can you separate this out into a separate PR please

sr55 replied:
Nevermind :)

sr55 commented on contrib/libdav1d/A00-add-cross-file.patch:
```diff
@@ -0,0 +1,19 @@
+diff -Naur /dev/null dav1d-f1b756ef5bdc0bb759b2b140f15362fec024c1ff/x86_64-w64-mingw32.meson
```
Been submitted upstream?

f3ndot replied:
Not yet, but happy to. @EwoutH is there a reason why I shouldn't submit this as a PR upstream to dav1d?

f3ndot replied:
Upstream merge request: https://code.videolan.org/videolan/dav1d/merge_requests/791

sr55 commented:
Looks fine on the surface. I'll find some time to test maybe tomorrow and we can look to get this in.

sr55 commented:
Am I being daft, or is https://github.com/HandBrake/HandBrake/pull/1864.patch including stuff from older revisions?  I.e. why is travis there when the head change here doesn't have anything with travis in it?

f3ndot commented:
@sr55 it's entirely possible I mucked up the merges because I accidentally included one from my experimental `f3ndot/master`.

I'll try and get this fixed tonight with some smart reverts and force-pushes

f3ndot commented:
@sr55 ok, after some git surgery everything is looking cleaned up and proper now. Resetted away troublesome merges, dropped commits no longer applicable (the Travis CI ones), and rebased off of latest master. Checking the `.patch` file everything looks cromulent.

sr55 commented:
ERROR: dav1d >= 0.2.1 not found using pkg-config  

building ffmpeg.

libdva1d appears to have compiled. 

Menson is a bit of a pest :(.  Ubuntu 18.10 or later has a new enough version.  18.04 needs an upgrade as it only has 0.45

MinGW Build.

Nomis101 commented:
Maybe it would also be useful than to add ninja and meson to the mac-toolchain-build script. It took me half an hour to find out how to install meson on macOS.

bradleysepos commented:
Sure, I can work on that once this goes in. Any tips, @Nomis101?

```bash
bradley$ which meson
/Library/Frameworks/Python.framework/Versions/3.7/bin/meson
```

Nomis101 commented:
> ERROR: dav1d >= 0.2.1 not found using pkg-config

I see this error only, if I use 7f1d089 alone. If I build with 7f1d089 and f3d9b69 I see the following:

```
 : Undefined symbols for architecture x86_64:
  :   "_dav1d_close", referenced from:
  :       _libdav1d_close in libavcodec.a(libdav1d.o)
  :   "_dav1d_data_unref", referenced from:
  :       _libdav1d_close in libavcodec.a(libdav1d.o)
  :       _libdav1d_flush in libavcodec.a(libdav1d.o)
  :   "_dav1d_data_wrap", referenced from:
  :       _libdav1d_receive_frame in libavcodec.a(libdav1d.o)
  :   "_dav1d_default_settings", referenced from:
  :       _libdav1d_init in libavcodec.a(libdav1d.o)
  :   "_dav1d_flush", referenced from:
  :       _libdav1d_flush in libavcodec.a(libdav1d.o)
  :   "_dav1d_get_picture", referenced from:
  :       _libdav1d_receive_frame in libavcodec.a(libdav1d.o)
  :   "_dav1d_open", referenced from:
  :       _libdav1d_init in libavcodec.a(libdav1d.o)
  :   "_dav1d_picture_unref", referenced from:
  :       _libdav1d_receive_frame in libavcodec.a(libdav1d.o)
  :   "_dav1d_send_data", referenced from:
  :       _libdav1d_receive_frame in libavcodec.a(libdav1d.o)
  :   "_dav1d_version", referenced from:
  :       _libdav1d_init in libavcodec.a(libdav1d.o)
  : ld: symbol(s) not found for architecture x86_64
  : clang: error: linker command failed with exit code 1 (use -v to see invocation)
  : 
  : ** BUILD FAILED **
  : 
  : 
  : The following build commands failed:
  : 	Ld /Volumes/Developer/HandBrake/build/xroot/HandBrakeCLI normal x86_64
  : (1 failure)
  : make: *** [macosx.build] Error 65
```

This is on macOS 10.14.


> Sure, I can work on that once this goes in. Any tips, @Nomis101?

Not yet.

galad87 commented:
If you are building in Xcode or from the cli with Xcode enabled you need to add the new library to the link phase in the Xcode project too.

Nomis101 commented:
With adding the new library to the link phase in the Xcode project, it builds and works nicely for me on Mac. [Like this](https://github.com/HandBrake/HandBrake/files/3570330/patch.txt).

Nomis101 commented:
> Sure, I can work on that once this goes in. Any tips, @Nomis101?

Something like this works for me. Meson needs the Python.framework to be installed. Haven't tested what will happen if it is not installed.

```
diff --git a/scripts/mac-toolchain-build b/scripts/mac-toolchain-build
index 6b4a16bad..a2fd2d280 100755
--- a/scripts/mac-toolchain-build
+++ b/scripts/mac-toolchain-build
@@ -34,7 +34,7 @@ function mac_toolchain_build {
         MAKEJOBS="4"
     fi
     SUDO=
-    TOTAL=6
+    TOTAL=8
 
     # functions
     function print_fail_and_exit {
@@ -89,6 +89,12 @@ function mac_toolchain_build {
     printf "Downloading [%02i/%02i] %s " "6" "${TOTAL}" "nasm 2.14.02"
     curl -Lf --connect-timeout 30 https://www.nasm.us/pub/nasm/releasebuilds/2.14.02/nasm-2.14.02.tar.bz2 -o "nasm-2.14.02.tar.bz2" >/dev/null 2>&1 || print_fail_and_exit
     echo -en "${CREL}"
+    printf "Downloading [%02i/%02i] %s " "7" "${TOTAL}" "ninja 1.9.0"
+    curl -Lf --connect-timeout 30 https://github.com/ninja-build/ninja/archive/v1.9.0.tar.gz -o "ninja-1.9.0.tar.gz" >/dev/null 2>&1 || print_fail_and_exit
+    echo -en "${CREL}"
+    printf "Downloading [%02i/%02i] %s " "8" "${TOTAL}" "meson 0.51.2"
+    curl -Lf --connect-timeout 30 https://github.com/mesonbuild/meson/archive/0.51.2.tar.gz -o "meson-0.51.2.tar.gz" >/dev/null 2>&1 || print_fail_and_exit
+    echo -en "${CREL}"
     printf "Downloading [%02i/%02i] complete.\n" "${TOTAL}" "${TOTAL}"
 
     # autoconf
@@ -157,6 +163,25 @@ function mac_toolchain_build {
     ${SUDO} make install >>../nasm-2.14.02.log 2>&1 || print_fail_and_exit
     echo -en "${CREL}"
 
+    # ninja
+    cd "${TEMP_DIR}"
+    printf "Building    [%02i/%02i] %s " "7" "${TOTAL}" "ninja 1.9.0"
+    [[ "${SUDO}" != "" ]] && ${SUDO} -v
+    tar -xf ninja-1.9.0.tar.gz >/dev/null 2>&1 || print_fail_and_exit
+    cd ninja-1.9.0 >/dev/null 2>&1 || print_fail_and_exit
+    ./configure.py --bootstrap >../ninja-1.9.0.log 2>&1 || print_fail_and_exit
+    ${SUDO} mv ninja /usr/local/bin >>../ninja-1.9.0.log 2>&1 || print_fail_and_exit
+    echo -en "${CREL}"
+
+    # meson
+    cd "${TEMP_DIR}"
+    printf "Building    [%02i/%02i] %s " "8" "${TOTAL}" "meson 0.51.2"
+    [[ "${SUDO}" != "" ]] && ${SUDO} -v
+    tar -xf meson-0.51.2.tar.gz >/dev/null 2>&1 || print_fail_and_exit
+    cd meson-0.51.2 >/dev/null 2>&1 || print_fail_and_exit
+    ${SUDO} python3 setup.py install >>../meson-0.51.2.log 2>&1 || print_fail_and_exit
+    echo -en "${CREL}"
+
     # done
     printf "Building    [%02i/%02i] complete.\n" "${TOTAL}" "${TOTAL}"
     rm -rf "${TEMP_DIR}"

```

f3ndot commented:
> Menson is a bit of a pest :(. Ubuntu 18.10 or later has a new enough version. 18.04 needs an upgrade as it only has 0.45

Yeah, [per their own docs](https://mesonbuild.com/Getting-meson.html), it looks like using `pip` is the preferred method for getting meson

> ERROR: dav1d >= 0.2.1 not found using pkg-config
> 
> building ffmpeg.
> 
> libdva1d appears to have compiled.

Ugh. I'll spend some time debugging this this weekend and hopefully get this merge-ready. Complete with mac toolchain changes as suggested above

f3ndot commented:
Was able to get a successful compile once release buildtype was set for dav1d with 

```bash
./configure --launch-jobs=8 --disable-gtk --launch --force 
```

Running on my Window's Ubuntu 18.04 Bionic subsystem (`Linux JUSTIN 4.4.0-17763-Microsoft #379-Microsoft Wed Mar 06 19:16:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux`)

Confirmed with a AV1 test file that everything worked A-OK for linux.

However, cross-compilation for Windows failed reproducing the same issue/error that @sr55 cited. My assumption is my cross-compilation patch for meson is insufficient. More research is required by me. Advice is absolutely welcome.

Nomis101 commented:
> However, cross-compilation for Windows failed reproducing the same issue/error that @sr55 cited. My assumption is my cross-compilation patch for meson is insufficient. More research is required by me. Advice is absolutely welcome.

Are you cross-compiling from Linux or from macOS for Windows? I've tried cross-compiling from macOS for Windows (x86_64-w64-mingw32), this worked nicely, no errors.

f3ndot commented:
@Nomis101 in https://github.com/HandBrake/HandBrake/commit/07f1ae570ea30d521904ff8a080deae279eeb6e6 I fixed the actual cross-compile file inclusion conditional. Compiles fine for me now

f3ndot commented:
@sr55 care to give this another go?

sr55 commented:
Looks better now. Compiled fine

sr55 commented:
@jstebbins @bradleysepos  @galad87  if you guys are happy, shall we merge?

Nomis101 commented:
This is currently not building on Mac with Xcode, see galad87's comment above
https://github.com/HandBrake/HandBrake/pull/1864#issuecomment-527178518
But this is quite easy to fix.

bradleysepos commented:
Haven't had a chance to test here, but if others are satisfied, I have no objection.

f3ndot committed:
scripts: Include meson and ninja in mac toolchain.

bradleysepos commented:
Thanks @f3ndot for 6349a2b, will test here shortly. Been a bit behind as of late.

bradleysepos commented:
meson failed with permissions issues, will investigate:

```
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/meson.egg-info/PKG-INFO: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/meson.egg-info/SOURCES.txt: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/meson.egg-info/entry_points.txt: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/meson.egg-info/top_level.txt: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/meson.egg-info/dependency_links.txt: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/meson.egg-info: Directory not empty
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/dist/meson-0.51.2-py3.7.egg: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/dist: Directory not empty
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild/__pycache__/coredata.cpython-37.pyc: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild/__pycache__/mesonlib.cpython-37.pyc: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild/__pycache__/mlog.cpython-37.pyc: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild/__pycache__/__init__.cpython-37.pyc: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild/__pycache__: Directory not empty
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild/wrap/__pycache__/__init__.cpython-37.pyc: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild/wrap/__pycache__: Directory not empty
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild/wrap: Directory not empty
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/mesonbuild: Directory not empty
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/bdist.macosx-10.9-x86_64: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/build.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/minit.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/coredata.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/cmake/client.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/cmake/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/cmake/interpreter.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/cmake/common.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/cmake: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/rewriter.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/mesonlib.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/optinterpreter.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/misc.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/ui.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/boost.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/platform.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/dev.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/data/CMakeListsLLVM.txt: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/data/CMakePathInfo.txt: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/data/CMakeLists.txt: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/data: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies/base.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/dependencies: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/minstall.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend/vs2015backend.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend/xcodebackend.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend/vs2019backend.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend/backends.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend/vs2017backend.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend/vs2010backend.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend/ninjabackend.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/backend: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/mparser.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/mlog.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/wrap/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/wrap/wraptool.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/wrap/wrap.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/wrap: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/interpreter.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/msubprojects.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/mconf.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/msetup.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/mesonmain.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/envconfig.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/regen_checker.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/gettext.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/clangformat.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/symbolextractor.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/coverage.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/scanbuild.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/yelphelper.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/depfixer.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/cleantrees.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/commandrunner.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/hotdochelper.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/uninstall.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/dirchanger.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/vcstagger.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/meson_exe.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/msgfmthelper.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/gtkdochelper.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/dist.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts/delwithsuffix.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/scripts: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/environment.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/ast/postprocess.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/ast/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/ast/interpreter.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/ast/visitor.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/ast/introspection.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/ast/printer.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/ast: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/mtest.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/munstable_coredata.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/java.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/objcpp.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/d.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/cpp.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/vala.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/cs.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/c.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/cuda.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/clike.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/objc.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/swift.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/compilers.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/c_function_attributes.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/fortran.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers/rust.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/compilers: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/python3.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/unstable_cuda.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/dlang.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/cmake.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/qt4.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/qt.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/qt5.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/__init__.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/pkgconfig.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/gnome.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/python.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/unstable_kconfig.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/sourceset.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/i18n.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/windows.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/rpm.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/modtest.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/hotdoc.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/unstable_icestorm.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules/unstable_simd.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/modules: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/interpreterbase.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/linkers.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild/mintro.py: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib/mesonbuild: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build/lib: Permission denied
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2/build: Directory not empty
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr/meson-0.51.2: Directory not empty
rm: /var/folders/z2/qsf1k22d1c990_ryjc0vvpqr0000gn/T/toolchain-NUaLCr: Directory not empty
```

bradleysepos commented on scripts/mac-toolchain-build:
```diff
@@ -157,6 +163,25 @@ function mac_toolchain_build {
+
     # done
     printf "Building    [%02i/%02i] complete.\n" "${TOTAL}" "${TOTAL}"
     rm -rf "${TEMP_DIR}"
```
I believe we'll need to add `${SUDO} ` here to allow removing meson's temp directory properly.

f3ndot committed:
macosx: Add libdav1d.a to Xcode build settings.

f3ndot commented:
@bradleysepos no worries. @Nomis101 I wasn't able to get a successful build with your XCode patch, but doing it myself through their GUI I was able to get a successful build.

Nomis101 commented:
> @Nomis101 I wasn't able to get a successful build with your XCode patch, but doing it myself through their GUI I was able to get a successful build.

Interesting, for me it is working nicely. What error did you see?

In this line https://github.com/HandBrake/HandBrake/pull/1864/commits/e8201f8d9b2ca38eb06a63cba1fa7b73eddb8c39#diff-abd6149eef9f351173644bba1dd182ffR806
The path needs to be `path = external/contrib/lib/libdav1d.a`. Similar to all the other `*.a` references in project.pbxproj.
https://github.com/HandBrake/HandBrake/blob/master/macosx/HandBrake.xcodeproj/project.pbxproj#L392

bradleysepos commented:
After applying @Nomis101's and my suggested changes, HandBrake builds and decodes AV1 on macOS.

bradleysepos commented:
I will fix and merge.

bradleysepos commented:
Merged with minor cleanup. Thanks all!

bradleysepos commented:
Addendum, build failure on Arch Linux: https://pastebin.com/iNUgzVkL

Looks like we might need to disable building additional dav1d tools.

bradleysepos commented:
Fixed in ab83daf27549299b4869dcfdfdf1f8a25da2a834.

f3ndot commented:
Awesome thanks @bradleysepos! As you can no doubt tell I'm a bit of a novice in this space.

bradleysepos commented:
You did quite well. Thanks very much for your contributions. 😸

EwoutH commented:
Thank you both on a great job!
